File: mobile-client/ios/App/Access/MobilingAppShell.swift
import SwiftUI

public struct MobilingAppShell: View {
    @State private var currentScreen: AccessScreen = .welcome
    private let authFeatureBridge: AuthFeatureBridge?

    public init(authFeatureBridge: AuthFeatureBridge? = nil) {
        self.authFeatureBridge = authFeatureBridge
    }

    public var body: some View {
        Group {
            switch currentScreen {
            case .welcome:
                AccessWelcomeView(
                    onSignIn: { currentScreen = .signIn },
                    onCreateAccess: { currentScreen = .register }
                )
            case .signIn:
                SignInView(
                    onBack: { currentScreen = .welcome },
                    onCreateAccess: { currentScreen = .register },
                    onStartAccess: { request in
                        guard let authFeatureBridge else {
                            return nil
                        }
                        return try await authFeatureBridge.start(request: request)
                    },
                    onAccessSession: { payload in currentScreen = payload.toAccessScreen() }
                )
            case .register:
                RegisterAccessView(
                    onBack: { currentScreen = .welcome },
                    onSignIn: { currentScreen = .signIn },
                    onRegisterAccess: { request in
                        guard let authFeatureBridge else {
                            return nil
                        }
                        return try await authFeatureBridge.register(request: request)
                    },
                    onAccessSession: { payload in currentScreen = payload.toAccessScreen() }
                )
            case .verificationRequired:
                VerificationRequiredView(
                    onBack: { currentScreen = .signIn },
                    onUseDifferentAccess: { currentScreen = .welcome }
                )
            case .secondFactorRequired:
                SecondFactorRequiredView(
                    onBack: { currentScreen = .signIn },
                    onUseDifferentAccess: { currentScreen = .welcome }
                )
            }
        }
    }
}

File: mobile-client/android/app/src/main/java/app/mobiling/client/access/MobilingAppShell.kt
package app.mobiling.client.access

import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import app.mobiling.client.auth.AuthFeatureBridge

@Composable
fun MobilingAppShell(
    authFeatureBridge: AuthFeatureBridge? = null,
) {
    var currentScreen by rememberSaveable { mutableStateOf(AccessScreen.Welcome) }

    Surface(Modifier.fillMaxSize()) {
        when (currentScreen) {
            AccessScreen.Welcome -> AccessWelcomeScreen(
                onSignIn = { currentScreen = AccessScreen.SignIn },
                onCreateAccess = { currentScreen = AccessScreen.Register },
            )

            AccessScreen.SignIn -> SignInScreen(
                onBack = { currentScreen = AccessScreen.Welcome },
                onCreateAccess = { currentScreen = AccessScreen.Register },
                onStartAccess = { request -> authFeatureBridge?.start(request) },
                onAccessSession = { payload -> currentScreen = payload.toAccessScreen() },
            )

            AccessScreen.Register -> RegisterAccessScreen(
                onBack = { currentScreen = AccessScreen.Welcome },
                onSignIn = { currentScreen = AccessScreen.SignIn },
                onRegisterAccess = { request -> authFeatureBridge?.register(request) },
                onAccessSession = { payload -> currentScreen = payload.toAccessScreen() },
            )

            AccessScreen.VerificationRequired -> VerificationRequiredScreen(
                onBack = { currentScreen = AccessScreen.SignIn },
                onUseDifferentAccess = { currentScreen = AccessScreen.Welcome },
            )

            AccessScreen.SecondFactorRequired -> SecondFactorRequiredScreen(
                onBack = { currentScreen = AccessScreen.SignIn },
                onUseDifferentAccess = { currentScreen = AccessScreen.Welcome },
            )
        }
    }
}
